File: extension/substrate/prompts.ts
// The TS plane's prompt render seam — the twin of perk/prompts.py::render.
//
// Templates are loaded by explicit `name` (root-relative under `prompts/`) via promptsDir(),
// the same directory the Python twin reads. The feature surface is the frozen mini-jinja subset
// (`shared/contracts.md §8.31`): `{{ var }}` substitution, `{% include %}`, and
// `{% if %}`/`{% elif %}`/`{% else %}`/`{% endif %}` conditionals. A missing (or non-string)
// variable fails loudly rather than rendering an empty string.
//
// jinja2 is the REFERENCE engine: the committed golden bytes under prompts/_fixtures/golden/
// ARE jinja2's output, and this seam must reproduce them byte-for-byte (enforced by prompts.test.ts
// + tests/test_prompts.py). Rendering is delegated to the vendored, zero-dependency
// ./miniJinja.ts renderer — which bakes in the frozen render config (trim_blocks on so a block
// tag on its own line emits no spurious newline, lstrip off, trailing newline preserved) and
// owns the filesystem (resolving `name` and every `{% include %}` under promptsDir()). Vendoring
// keeps the extension zero-runtime-dep / loadable from a bare git clone (guarded by
// extension/bareImportGuard.test.ts).
//
// This seam is LIVE in production: render is imported by the worker, the warm doors, the
// factories, and the mode/adapter context modules (the `prompts/contexts/` injections).
//
// Besides the render seam itself, the module carries ONE cross-plane selector wrapping it:
// `planReadInstruction`, the per-backend plan-read arm shared by the doors and the
// stage-execution seam.

import { render as miniJinjaRender } from "./miniJinja.ts";

/** Render the template at `name` (root-relative under `prompts/`) with `vars`. */
export function render(name: string, vars: Record<string, unknown>): string {
  return miniJinjaRender(name, vars);
}

/**
 * The per-backend plan-read instruction — the prompt SSOT for "how do I read the saved
 * plan". Byte-identical to `src/perk/run/launch/prompts.py::_plan_read_instruction` (the Python
 * twin — likewise a prompts module); drift in either plane fails the paired parity suites.
 * `github` reads via `gh`; `linear` points at the pi-mono-linear tools with an `open <url>`
 * fallback; any other provider falls back to opening the url.
 *
 * The wording lives in the canonical templates `prompts/common/plan-read/*.md`, rendered
 * identically by both planes via the shared render seam (contracts.md §8.31); branching stays in
 * code — only the arm chosen and the vars passed differ. Golden-fixture parity (the three
 * `plan-read-*` cases) plus a thin per-arm selection test replace the dedicated substring parity.
 */
export function planReadInstruction(provider: string, prId: string, url: string): string {
  if (provider === "github") return render("common/plan-read/github.md", { pr_id: prId, url });
  if (provider === "linear") return render("common/plan-read/linear.md", { pr_id: prId, url });
  return render("common/plan-read/other.md", { pr_id: prId, url });
}

File: perk/prompts.py
"""The Python plane's prompt render seam (TS twin: ``extension/substrate/prompts.ts``).

Templates are loaded by explicit ``name`` (root-relative under ``prompts/``) via
:func:`perk._resources.prompts_dir`, the same directory the TS twin reads. The feature
surface is intentionally small — ``{{ var }}`` substitution and ``{% include %}`` — and
``StrictUndefined`` makes a missing variable fail loudly rather than render an empty string.

jinja2 is the **reference engine**: the committed golden bytes under
``prompts/_fixtures/golden/`` ARE this seam's output, and the TS twin's vendored mini-jinja
renderer (``extension/substrate/miniJinja.ts``) must reproduce them byte-for-byte. Golden parity
is enforced by ``tests/test_prompts.py`` + ``extension/substrate/prompts.test.ts``. The Environment
config below is the parity baseline both engines share (autoescape off, ``trim_blocks`` on so a
block tag on its own line emits no spurious newline — letting conditional templates keep their tags
off the content lines while preserving indentation — ``lstrip_blocks`` off, and
``keep_trailing_newline`` on so jinja2 does not strip a trailing ``\\n`` the TS renderer keeps).
"""

from collections.abc import Mapping

from jinja2 import Environment, FileSystemLoader, StrictUndefined

from perk._resources import prompts_dir

_env = Environment(
    loader=FileSystemLoader(str(prompts_dir())),
    undefined=StrictUndefined,
    autoescape=False,
    keep_trailing_newline=True,
    trim_blocks=True,
    lstrip_blocks=False,
)


def render(name: str, variables: Mapping[str, object]) -> str:
    """Render the template at ``name`` (root-relative under ``prompts/``) with ``variables``.

    The render contract is **string-only** (the frozen mini-jinja subset, ``contracts.md §8.31``):
    every variable value must be a ``str``. This mechanically enforces the same contract the TS
    twin's vendored renderer enforces — the difference being only *when* it fires: the TS renderer
    throws lazily on a referenced non-string, this validates the whole var map eagerly. Both forbid
    the silent ``str(value)`` coercion (a future ``False`` must never render ``"False"``).
    """
    for key, value in variables.items():
        if not isinstance(value, str):
            raise TypeError(
                f"perk prompts: variable {key!r} is {type(value).__name__}, not str "
                "(the render contract is string-only)"
            )
    return _env.get_template(name).render(variables)
